julia/fastsum/libfastsumjulia.c

```diff
@@ -0,0 +1,146 @@
+	C (*kernel)(R, int, const R *);
+	
+	if ( strcmp(s, "gaussian") == 0 )
+		kernel = gaussian;
+	else if ( strcmp(s, "multiquadric") == 0 )
+		kernel = multiquadric;
+	else if ( strcmp(s, "inverse_multiquadric") == 0 )
+		kernel = inverse_multiquadric;
+	else if ( strcmp(s, "logarithm") == 0 )
+		kernel = logarithm;
+	else if ( strcmp(s, "thinplate_spline") == 0 )
+		kernel = thinplate_spline;
+	else if ( strcmp(s, "one_over_square") == 0 )
+		kernel = one_over_square;
+	else if ( strcmp(s, "one_over_modulus") == 0 )
+		kernel = one_over_modulus;
+	else if ( strcmp(s, "one_over_x") == 0 )
+		kernel = one_over_x;
+	else if ( strcmp(s, "inverse_multiquadric3") == 0 )
+		kernel = inverse_multiquadric3;
+	else if ( strcmp(s, "sinc_kernel") == 0 )
+		kernel = sinc_kernel;
+	else if ( strcmp(s, "cosc") == 0 )
+		kernel = cosc;
+	else if ( strcmp(s, "cot") == 0 )
+		kernel = kcot;
+	else if ( strcmp(s, "one_over_cube") == 0 )
+		kernel = one_over_cube;
+	else if ( strcmp(s, "log_sin") == 0 )
+		kernel = log_sin;
+	else if ( strcmp(s, "laplacian_rbf") == 0 )
+		kernel = laplacian_rbf;
+	else {
+		s = "multiquadric";
+		kernel = multiquadric;
+	}
+	
+	fastsum_init_guru_kernel( p, d, kernel, c, f | STORE_PERMUTATION_X_ALPHA, n, ps, eps_I, eps_B);
+	p -> x = 0;
+	p -> y = 0;
+	
+}
+
+double* jfastsum_set_x( fastsum_plan* p, double* x, int N, int nn, int m ){	
+	
+	if ( !(p->x) ) 
+		fastsum_init_guru_source_nodes( p, N, nn, m );
+	else {
+		fastsum_finalize_source_nodes( p );
+		fastsum_init_guru_source_nodes( p, N, nn, m );
+	}	
+	
+	int d = p -> d;
+	
+	if ( p -> permutation_x_alpha == NULL ) {
+	
+		for ( int k = 0; k < N; k++ )
+			for ( int t = 0; t < d; t++)
+				p -> x[k*d+t] = x[k+t*N];
+			
+	} else {
+		
+		for ( int k = 0; k < N; k++ )
+			for ( int t = 0; t < d; t++)
+				p -> x[k*d+t] = x[p->permutation_x_alpha[k]+t*N];
+		
+	}
+	
+	fastsum_precompute_source_nodes( p );
+		
+	return p -> x;
+}
+
+double* jfastsum_set_y( fastsum_plan* p, double* y, int M, int nn, int m ){
+	if ( !(p->y) ) 
+		fastsum_init_guru_target_nodes( p, M, nn, m );
+	else {
+		fastsum_finalize_target_nodes( p );
+		fastsum_init_guru_target_nodes( p, M, nn, m );
+	}	
+	
+	int d = p -> d;
+	
+	for ( int j = 0; j < M; j++ )
+		for ( int t = 0; t < d; t++ )
+			p -> y[j*d+t] = y[j+t*M];
+		
+	fastsum_precompute_target_nodes( p );
+		
+	return p -> y;
+}
+
+double _Complex* jfastsum_set_alpha( fastsum_plan* p, double _Complex* alpha ){
+	int N = p -> N_total;
+	
+	for ( int k = 0; k < N; k++ )
+		if ( p -> permutation_x_alpha == NULL )
+			p -> alpha[k] = alpha[k];
+		else
+			p -> alpha[k] = alpha[p->permutation_x_alpha[k]];
+			
+	return p -> alpha;
+}
+
+
+double _Complex* jfastsum_trafo( fastsum_plan* p ){
+	fastsum_trafo( p );
+	return p -> f;
+}
+
+double _Complex* jfastsum_exact( fastsum_plan* p ){
+	fastsum_exact( p );
+	return p -> f;
+}
+
+void jfastsum_finalize( fastsum_plan* p ){
+	fastsum_finalize_source_nodes( p );
```

Comment:
we should check if the source nodes were initialized before calling fastsum_finalize_source_nodes

Reply:
This won't be necessary with the new structure since the finalizer in julia checks if the plan was initialized.